tooling/templatize/pkg/pipeline/run.go

```diff
@@ -78,56 +82,320 @@ func (o ShellOutput) GetValue(_ string) (*OutPutValue, error) {
 // Outputs stores output values indexed by resource group name and step name.
 type Outputs map[string]map[string]Output
 
-func RunPipeline(pipeline *types.Pipeline, ctx context.Context, options *PipelineRunOptions) (Outputs, error) {
-	outPuts := make(Outputs)
+type Executor func(s types.Step, ctx context.Context, executionTarget ExecutionTarget, options *PipelineRunOptions, outPuts Outputs) (Output, error)
+
+func RunPipeline(pipeline *types.Pipeline, ctx context.Context, options *PipelineRunOptions, executor Executor) (Outputs, error) {
+	logger, err := logr.FromContext(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	logger.Info("Generating execution graph.")
+	nodes, err := createExecutionGraph(ctx, pipeline, options.Region, options.SubsciptionLookupFunc)
+	if err != nil {
+		return nil, err
+	}
+
+	lock := &sync.RWMutex{}
+	executed := sets.Set[types.StepDependency]{}
+	queued := sets.Set[types.StepDependency]{}
+	outputs := make(Outputs)
```

Comment:
I think a struct containing these items so the lock access is closely associated and passed a single unit will likely make life easier.